Source/ProgressionSystemRuntime/Private/Widgets/PSOverlayWidget.cpp

```diff
@@ -4,11 +4,113 @@
 #include "Widgets/PSOverlayWidget.h"
 
 #include "Components/Image.h"
+#include "Curves/CurveFloat.h"
+#include "Data/PSDataAsset.h"
 
 #include UE_INLINE_GENERATED_CPP_BY_NAME(PSOverlayWidget)
 
-// Sets the visibility of the background overlay and lock icon.
-void UPSOverlayWidget::SetOverlayVisibility(ESlateVisibility VisibilitySlate)
+// Sets the visibility of the overlay elements and playing fade animation if needed
+void UPSOverlayWidget::SetOverlayVisibility(ESlateVisibility VisibilitySlate, bool bShouldPlayFadeAnimation/* = false*/)
+{
+	if (!bShouldPlayFadeAnimation)
+	{
+		SetOverlayItemsVisibility(VisibilitySlate);
+		return;
+	}
+
+	bShouldPlayFadeAnimationInternal = bShouldPlayFadeAnimation;
+
+	if (VisibilitySlate == ESlateVisibility::Visible)
+	{
+		ESlateVisibility PrevOverlayOpacity = PSCBackgroundOverlay->GetVisibility();
+		ESlateVisibility PrevIconOpacity = PSCBackgroundIconLock->GetVisibility();
+
+		//if new visibility is same as previous animation is not required 
+		if (PrevOverlayOpacity == VisibilitySlate && PrevIconOpacity == VisibilitySlate)
+		{
+			bShouldPlayFadeAnimationInternal = false;
+		}
+		bIsFadeInAnimationInternal = true;
+		SetOverlayItemsVisibility(VisibilitySlate);
+	}
+	else
+	{
+		bIsFadeInAnimationInternal = false;
+	}
+
+	const UWorld* World = GetWorld();
+	if (!ensureMsgf(World, TEXT("ASSERT: [%i] %s:\n'World' is not valid!"), __LINE__, *FString(__FUNCTION__)))
+	{
+		return;
+	}
+
+	StartTimeFadeAnimationInternal = World->GetTimeSeconds();
+}
+
+// overrides NativeTick to make the user widget tickable
+void UPSOverlayWidget::NativeTick(const FGeometry& MyGeometry, float InDeltaTime)
+{
+	Super::NativeTick(MyGeometry, InDeltaTime);
+	if (bShouldPlayFadeAnimationInternal)
+	{
+		const bool bIsFinished = !FadeOverlayElementsAnimation(StartTimeFadeAnimationInternal);
+		if (bIsFinished)
+		{
+			bShouldPlayFadeAnimationInternal = false;
+			StartTimeFadeAnimationInternal = 0.f;
+			if (!bIsFadeInAnimationInternal)
+			{
+				SetOverlayItemsVisibility(ESlateVisibility::Collapsed);
+			}
+		}
+	}
+}
+
+// Event to execute when widget is ready
+void UPSOverlayWidget::NativeConstruct()
+{
+	Super::NativeConstruct();
+	FadeCurveFloatInternal = UPSDataAsset::Get().GetFadeCurveFloat();
+}
+
+// Play the overlay elemetns fade-in/fade-out animation. Uses the internal FadeCurveFloatInternal initialized in NativeConstruct
+bool UPSOverlayWidget::FadeOverlayElementsAnimation(float& StartTimeRef)
+{
+	if (!StartTimeRef || !FadeCurveFloatInternal || !bShouldPlayFadeAnimationInternal)
+	{
+		StartTimeRef = 0.f;
+		return false;
+	}
+
+	const UWorld* World = GetWorld();
+	check(World);
```

Comment:
`check` is very risky here for World as it's called from Tick, so it might crash when reloading the level etc. Since we are in tick, please do silent return for invalid World, without any errors, as it's expected

Reply:
changed: https://github.com/h4rdmol/ProgressionSystem/pull/20/commits/31e73b01eab9b78b7fe3e1838687c7b51f76c3bc